File: js/src/lexer.ts
import { TokenError } from './errors';

export enum TokenType {
  Ident = 'IDENT',
  String = 'STRING',
  Number = 'NUMBER',
  Boolean = 'BOOLEAN',
  Null = 'NULL',
  LBrace = 'LBRACE',
  RBrace = 'RBRACE',
  LBracket = 'LBRACKET',
  RBracket = 'RBRACKET',
  Equals = 'EQUALS',
  Comma = 'COMMA',
  Dot = 'DOT',
  Dollar = 'DOLLAR',
  Newline = 'NEWLINE',
  Comment = 'COMMENT',
  Eof = 'EOF',
}

export interface Token {
  type: TokenType;
  literal: string;
  line: number;
  col: number;
}

export class Tokenizer {
  private input: string;
  private pos: number = 0;
  private line: number = 1;
  private col: number = 1;
  private tokens: Token[] = [];
  private peeked: Token | null = null;

  constructor(input: string) {
    this.input = input;
  }

  tokenize(): Token[] {
    while (this.pos < this.input.length) {
      this.skipWhitespace();
      if (this.pos >= this.input.length) break;

      const ch = this.input[this.pos];

      if (ch === '#') {
        this.scanComment();
      } else if (ch === '/' && this.peekChar() === '/') {
        this.scanLineComment();
      } else if (ch === '"' || ch === "'") {
        this.scanString();
      } else if (ch === '{') {
        this.addToken(TokenType.LBrace, '{');
        this.advance();
      } else if (ch === '}') {
        this.addToken(TokenType.RBrace, '}');
        this.advance();
      } else if (ch === '[') {
        this.addToken(TokenType.LBracket, '[');
        this.advance();
      } else if (ch === ']') {
        this.addToken(TokenType.RBracket, ']');
        this.advance();
      } else if (ch === '=') {
        this.addToken(TokenType.Equals, '=');
        this.advance();
      } else if (ch === ',') {
        this.addToken(TokenType.Comma, ',');
        this.advance();
      } else if (ch === '.') {
        this.addToken(TokenType.Dot, '.');
        this.advance();
      } else if (ch === '$') {
        this.addToken(TokenType.Dollar, '$');
        this.advance();
      } else if (ch === '~') {
        this.addToken(TokenType.Null, 'null');
        this.advance();
      } else if (ch === '-' || this.isDigit(ch)) {
        this.scanNumber();
      } else if (this.isIdentStart(ch)) {
        this.scanIdent();
      } else {
        throw new TokenError(
          `Unexpected character: '${ch}'`,
          this.line,
          this.col
        );
      }
    }

    this.addToken(TokenType.Eof, '');
    return this.tokens;
  }

  peek(): Token {
    if (this.peeked) return this.peeked;
    const saved = { pos: this.pos, line: this.line, col: this.col, tokens: this.tokens.length };
    const tokens = this.tokenize();
    this.peeked = tokens[tokens.length - 1] || { type: TokenType.Eof, literal: '', line: this.line, col: this.col };
    this.pos = saved.pos;
    this.line = saved.line;
    this.col = saved.col;
    this.tokens = this.tokens.slice(0, saved.tokens);
    return this.peeked;
  }

  private peekChar(): string {
    if (this.pos + 1 >= this.input.length) return '\0';
    return this.input[this.pos + 1];
  }

  private advance(): void {
    this.pos++;
    this.col++;
  }

  private addToken(type: TokenType, literal: string): void {
    this.tokens.push({ type, literal, line: this.line, col: this.col });
  }

  private skipWhitespace(): void {
    while (this.pos < this.input.length) {
      const ch = this.input[this.pos];
      if (ch === ' ' || ch === '\t' || ch === '\r') {
        this.advance();
      } else if (ch === '\n') {
        this.addToken(TokenType.Newline, '\n');
        this.pos++;
        this.line++;
        this.col = 1;
      } else {
        break;
      }
    }
  }

  private scanComment(): void {
    const start = this.pos;
    while (this.pos < this.input.length && this.input[this.pos] !== '\n') {
      this.pos++;
    }
    this.addToken(TokenType.Comment, this.input.substring(start, this.pos));
  }

  private scanLineComment(): void {
    const start = this.pos;
    this.pos += 2;
    this.col += 2;
    while (this.pos < this.input.length && this.input[this.pos] !== '\n') {
      this.pos++;
    }
    this.addToken(TokenType.Comment, this.input.substring(start, this.pos));
  }

  private scanString(): void {
    const quote = this.input[this.pos];
    const startLine = this.line;
    const startCol = this.col;
    this.advance();

    let result = '';
    while (this.pos < this.input.length) {
      const ch = this.input[this.pos];
      if (ch === '\n') {
        throw new TokenError(
          `Unterminated string at ${startLine}:${startCol}`,
          startLine,
          startCol
        );
      }
      if (ch === '\\' && this.pos + 1 < this.input.length) {
        this.advance();
        const escaped = this.input[this.pos];
        switch (escaped) {
          case 'n': result += '\n'; break;
          case 't': result += '\t'; break;
          case '\\': result += '\\'; break;
          case '"': result += '"'; break;
          case "'": result += "'"; break;
          default: result += '\\' + escaped; break;
        }
      } else if (ch === quote) {
        this.advance();
        this.addToken(TokenType.String, result);
        return;
      } else {
        result += ch;
      }
      this.advance();
    }
    throw new TokenError(
      `Unterminated string at ${startLine}:${startCol}`,
      startLine,
      startCol
    );
  }

  private scanNumber(): void {
    const start = this.pos;
    const startCol = this.col;
    if (this.input[this.pos] === '-') {
      this.advance();
    }
    while (this.pos < this.input.length && this.isDigit(this.input[this.pos])) {
      this.advance();
    }
    if (this.pos < this.input.length && this.input[this.pos] === '.') {
      this.advance();
      while (this.pos < this.input.length && this.isDigit(this.input[this.pos])) {
        this.advance();
      }
    }
    this.addToken(TokenType.Number, this.input.substring(start, this.pos));
  }

  private scanIdent(): void {
    const start = this.pos;
    while (this.pos < this.input.length && this.isIdentChar(this.input[this.pos])) {
      this.pos++;
      this.col++;
    }
    const word = this.input.substring(start, this.pos);
    switch (word) {
      case 'true':
      case 'false':
        this.addToken(TokenType.Boolean, word);
        break;
      case 'null':
        this.addToken(TokenType.Null, word);
        break;
      default:
        this.addToken(TokenType.Ident, word);
        break;
    }
  }

  private isDigit(ch: string): boolean {
    return ch >= '0' && ch <= '9';
  }

  private isIdentStart(ch: string): boolean {
    return (
      (ch >= 'a' && ch <= 'z') ||
      (ch >= 'A' && ch <= 'Z') ||
      ch === '_' ||
      ch === '-' ||
      ch === '.'
    );
  }

  private isIdentChar(ch: string): boolean {
    return this.isIdentStart(ch) || this.isDigit(ch);
  }
}

export function isIdent(s: string): boolean {
  if (s.length === 0) return false;
  const first = s[0];
  if (
    !((first >= 'a' && first <= 'z') || (first >= 'A' && first <= 'Z') || first === '_' || first === '-')
  ) {
    return false;
  }
  for (let i = 1; i < s.length; i++) {
    const ch = s[i];
    if (
      !(
        (ch >= 'a' && ch <= 'z') ||
        (ch >= 'A' && ch <= 'Z') ||
        (ch >= '0' && ch <= '9') ||
        ch === '_' ||
        ch === '-' ||
        ch === '.'
      )
    ) {
      return false;
    }
  }
  return true;
}

File: python/goml/lexer.py
"""Token definitions and tokenizer for GOML."""

from enum import Enum, auto
from typing import List, NamedTuple

from .errors import TokenError


class TokenType(Enum):
    IDENT = auto()
    STRING = auto()
    NUMBER = auto()
    BOOLEAN = auto()
    NULL = auto()
    LBRACE = auto()
    RBRACE = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    EQUALS = auto()
    COMMA = auto()
    DOT = auto()
    DOLLAR = auto()
    PIPE = auto()
    NEWLINE = auto()
    EOF = auto()


class Token(NamedTuple):
    type: TokenType
    value: object
    line: int
    col: int

    def __repr__(self):
        return f"Token({self.type.name}, {self.value!r}, line={self.line}, col={self.col})"


KEYWORDS = {
    "true": (TokenType.BOOLEAN, True),
    "false": (TokenType.BOOLEAN, False),
    "null": (TokenType.NULL, None),
    "~": (TokenType.NULL, None),
}

SINGLE_CHARS = {
    "{": TokenType.LBRACE,
    "}": TokenType.RBRACE,
    "[": TokenType.LBRACKET,
    "]": TokenType.RBRACKET,
    "=": TokenType.EQUALS,
    ",": TokenType.COMMA,
    ".": TokenType.DOT,
    "$": TokenType.DOLLAR,
    "|": TokenType.PIPE,
}


def tokenize(source: str) -> List[Token]:
    """Tokenize GOML source into a list of tokens."""
    tokens: List[Token] = []
    i = 0
    line = 1
    col = 1
    length = len(source)

    def advance() -> str:
        nonlocal i, line, col
        ch = source[i]
        i += 1
        if ch == "\n":
            line += 1
            col = 1
        else:
            col += 1
        return ch

    def peek(offset: int = 0) -> str:
        pos = i + offset
        if pos < length:
            return source[pos]
        return ""

    def skip_whitespace():
        nonlocal i, col
        while i < length and source[i] in " \t\r":
            if source[i] == "\t":
                col += 1
            else:
                col += 1
            i += 1

    def read_string(quote: str) -> str:
        nonlocal i, line, col
        start_line, start_col = line, col
        i += 1  # skip opening quote
        col += 1
        result: List[str] = []
        escape = False
        while i < length:
            ch = source[i]
            if escape:
                escape = False
                if ch == "n":
                    result.append("\n")
                elif ch == "t":
                    result.append("\t")
                elif ch == "\\":
                    result.append("\\")
                elif ch == quote:
                    result.append(quote)
                else:
                    result.append("\\" + ch)
                i += 1
                col += 1
            elif ch == "\\":
                escape = True
                i += 1
                col += 1
            elif ch == quote:
                i += 1
                col += 1
                return "".join(result)
            elif ch == "\n":
                line += 1
                col = 1
                i += 1
                result.append(ch)
            else:
                result.append(ch)
                i += 1
                col += 1
        raise TokenError("Unterminated string", start_line, start_col)

    def read_number() -> str:
        nonlocal i, col
        start = i
        if i < length and source[i] == "-":
            i += 1
            col += 1
        # Read digits
        while i < length and source[i].isdigit():
            i += 1
            col += 1
        # Read optional decimal part
        if i < length and source[i] == ".":
            # Peek ahead to see if this is part of a dotted identifier
            j = i + 1
            has_digit_after_dot = j < length and source[j].isdigit()
            if has_digit_after_dot:
                # Check if there are more dots after the decimal digits
                k = j + 1
                while k < length and source[k].isdigit():
                    k += 1
                if k < length and source[k] == ".":
                    # Multiple dots - this is an identifier like 192.168.1.1
                    # Return just the digits before the first dot
                    return source[start:i]
                # Single dot with digits - valid float
                i += 1
                col += 1
                while i < length and source[i].isdigit():
                    i += 1
                    col += 1
        return source[start:i]

    def read_ident() -> str:
        nonlocal i, col
        start = i
        while i < length and (source[i].isalnum() or source[i] in "_-"):
            i += 1
            col += 1
        return source[start:i]

    def read_ident_with_dots(prefix: str) -> str:
        nonlocal i, col
        result = prefix
        while i < length and source[i] == ".":
            result += source[i]
            i += 1
            col += 1
            # Read next part (digit or alpha)
            start_part = i
            while i < length and (source[i].isalnum() or source[i] in "_-"):
                i += 1
                col += 1
            result += source[start_part:i]
        return result

    def add_token(tok_type: TokenType, value: object):
        tokens.append(Token(tok_type, value, line, col))

    while i < length:
        ch = source[i]

        # Newlines
        if ch == "\n":
            advance()
            tokens.append(Token(TokenType.NEWLINE, "\n", line, col))
            continue

        # Skip whitespace (not newlines)
        if ch in " \t\r":
            skip_whitespace()
            continue

        # Comment
        if ch == "#":
            start_line, start_col = line, col
            while i < length and source[i] != "\n":
                advance()
            continue

        # String
        if ch == '"':
            start_line, start_col = line, col
            val = read_string('"')
            tokens.append(Token(TokenType.STRING, val, start_line, start_col))
            continue
        if ch == "'":
            start_line, start_col = line, col
            val = read_string("'")
            tokens.append(Token(TokenType.STRING, val, start_line, start_col))
            continue

        # Number (negative sign followed by digit)
        if ch == "-" and i + 1 < length and source[i + 1].isdigit():
            start_line, start_col = line, col
            num_str = read_number()
            # Check if this is actually part of a dotted identifier
            if i < length and source[i] == ".":
                # Read the rest as an identifier
                ident = read_ident_with_dots(num_str)
                tokens.append(Token(TokenType.IDENT, ident, start_line, start_col))
            elif "." in num_str:
                tokens.append(Token(TokenType.NUMBER, float(num_str), start_line, start_col))
            else:
                tokens.append(Token(TokenType.NUMBER, int(num_str), start_line, start_col))
            continue

        # Number
        if ch.isdigit():
            start_line, start_col = line, col
            num_str = read_number()
            # Check if this is actually part of a dotted identifier
            if i < length and source[i] == ".":
                # Read the rest as an identifier
                ident = read_ident_with_dots(num_str)
                tokens.append(Token(TokenType.IDENT, ident, start_line, start_col))
            elif "." in num_str:
                tokens.append(Token(TokenType.NUMBER, float(num_str), start_line, start_col))
            else:
                tokens.append(Token(TokenType.NUMBER, int(num_str), start_line, start_col))
            continue

        # Tilde for null
        if ch == "~":
            start_line, start_col = line, col
            advance()
            tokens.append(Token(TokenType.NULL, None, start_line, start_col))
            continue

        # Single character tokens
        if ch in SINGLE_CHARS:
            tok_type = SINGLE_CHARS[ch]
            advance()
            tokens.append(Token(tok_type, ch, line, col))
            continue

        # Identifiers / keywords
        if ch.isalpha() or ch == "_":
            start_line, start_col = line, col
            ident = read_ident()
            if ident in KEYWORDS:
                tok_type, val = KEYWORDS[ident]
                tokens.append(Token(tok_type, val, start_line, start_col))
            else:
                tokens.append(Token(TokenType.IDENT, ident, start_line, start_col))
            continue

        raise TokenError(f"Unexpected character: {ch!r}", line, col)

    tokens.append(Token(TokenType.EOF, None, line, col))
    return tokens
